### PR TITLE
RSDK-8905: Pigpio version update on bullseye

### PR DESCRIPTION
**Pull request by susmitaSanyal**
On bullseye, pigpio and all it's dependencies are running on version 1.79-1+rpt1. run.sh fails to install pigpio since dependencies are not met during installation. This PR address this issue and installs pigpio and libpigio-dev library in version 1.79-1+rpt1 in bullseye. bookworm doesn't have this issue.

### CODE REVIEW

**Review by randhid — APPROVED** [review]

**Comment by randhid** [comment]
link ticket number please

File: run.sh [hunk 1]
```diff
@@ -1,16 +1,34 @@
...
-systemctl enable pigpiod
+echo "Detected Debian version: $DEBIAN_VERSION"
 
-# start the pigpiod service
-echo "Starting pigpiod service..."
+# If Debian Bullseye, update specific versions of pigpio, libpigpio1, and libpigpio-dev
+if [ "$DEBIAN_VERSION" = "bullseye" ]; then
+    echo "Updating pigpio packages for Bullseye..."
+    apt-get update -qq
+    apt-get install -qqy pigpio=1.79-1+rpt1 libpigpio1=1.79-1+rpt1 libpigpio-dev=1.79-1+rpt1
```
[the remark this answers is not in the document]
Comment (Otterverse) [reply run.sh:11]: Lastly, why not just use this command for both distros. No harm in explicit versions being selected on bookworm too, is there?

**Review by Otterverse — CHANGES_REQUESTED** [review]
Minor things, but apt update is a slow thing that probably shouldn't run every start. Also, probably don't need the logic per distro, and can just run the same command (explicit version selection) on both.

File: run.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (susmitaSanyal) [reply run.sh:11]: yeah I agree with you. Verifying a few things, brb

[the remark this answers is not in the document]
Comment (susmitaSanyal) [reply run.sh]: checking for existing library version now instead of os version

File: run.sh [hunk 2]
```diff
@@ -1,7 +1,22 @@
 #!/bin/sh
 
-# install packages 
-apt-get install -qqy pigpio
+# if libpigpio is installed then we install pigpio 
+# with the same version of libpigpio
+if ! dpkg -s pigpio >/dev/null 2>&1; then 
+
+    # check if libpigpio is installed
+    if dpkg -s libpigpio-dev >/dev/nul 2>&1; then
```
[the remark this answers is not in the document]
Comment (Otterverse) [reply run.sh:8]:
```suggestion
    if dpkg -s libpigpio-dev >/dev/null 2>&1; then 
```

**Review by randhid — CHANGES_REQUESTED** [review]
This has changed enough from what I know worked that I'd like logs from the following tests, you can upload an rc candidate to the registry if it assists in testing, and I'd prefer that honestly.
- raspberry pi5 bookworm showing that it's skipping known packages.
- raspberry pi4 bullseye, use an older image from 2022 which is the timeframe I believe the pis with the incompatible prt1.1 package gets installed through the imager.

File: run.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (susmitaSanyal) [reply run.sh]: It's not, but what I've noticed is that when libpigpio-dev is installed then pigpio expects to the the same version of libpigppio-dev in the older versions of debien. 
When libpigpio-dev is not installed, `sudo apt-get install pigpio` installs whatever version depending on the OS. This piece is only necessary to support older raspberry-pis with older OS. We can most likely update it in the future

[the remark this answers is not in the document]
Comment (Otterverse) [reply run.sh]: If that's the concern, better to be explicit then. Check every package you actually need, if it's possible versions can be mismatched. Don't rely on a dependency chain (whose's requirements could change as updates come in) to incidentally do the right thing for your use case.

**Comment by susmitaSanyal** [comment]
@randhid @Otterverse ran into another instance of this problem where it was expecting a different version. At this point I think the best thing to do is just install the dependencies with specific version. We will only need to update the version number if there's a new version of pigpio dependencies. I am open to disagreement, but this seems like a simpler solution and thing to worry about.

File: run.sh [hunk 3]
```diff
@@ -1,7 +1,7 @@
 #!/bin/sh
 
-# install packages 
-apt-get install -qqy pigpio
+apt-get install -qqy libpigpio1=1.79-1+rpt1 libpigpio-dev=1.79-1+rpt1
```
Comment (Otterverse) [inline run.sh:3]: I'm fine with pinning versions here, but I really think it needs to include the versions of packages you actually use DIRECTLY. The module doesn't even use libpigpio1 or libpigpio-dev itself. It only depends DIRECTLY on libpigpiod-if2-1 (and requires pigpiod to be installed.) If you're going to force versions on only two packages, it should be these two. Everything is superfluous.

File: run.sh [hunk 4]
```diff
@@ -1,7 +1,7 @@
 #!/bin/sh
 
-# install packages 
-apt-get install -qqy pigpio
+apt-get install -qqy libpigpio1=1.79-1+rpt1 libpigpio-dev=1.79-1+rpt1 
+apt-get install pigpio
```
Comment (Otterverse) [inline run.sh:4]: Why is this on a separate line? Also, this package is the "transitional" package, so all it does is include ALL the other pigpio packages that have been split out. If this was causing conflicts, then add this (and pin the version as well) to the line above, so it's all one operation.

**Review by Otterverse — CHANGES_REQUESTED** [review]
Perhaps I'm unclear on the problems you've been encountering, so if my inline comments go against something you're doing for a very particular reason, let me know.

**Comment by randhid** [comment]
@Otterverse just for more context - the issue we've been seeing is that some images of raspian bullseye seem to install `libpigpio1=1.79-1+rpt1.1 libpigpio-dev=1.79-1+rpt1.1`, which seem to not adhere to what libgpiod1 is compatible with. 
```sh
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 pigpio : Depends: libpigpio1 (= 1.79-1+rpt1) but 1.79-1+rpt1.1 is to be installed
          Depends: libpigpio-dev (= 1.79-1+rpt1) but 1.79-1+rpt1.1 is to be installed
E: Unable to correct problems, you have held broken packages.
```

This was an issue fixed for bookworm with pigpio, which Susmita has more context on, but I could not get the module to run without erroring on bullseye with the dynamic link without forcing `rpt1` on some bullseye raspberry pi4s with `sudo apt install libpigpio1=1.79-1+rpt1 libpigpio-dev=1.79-1+rpt1`. This has popped up on 3 separate raspberry pis since the module's release, and I'm sure there are more.

**Comment by Otterverse** [comment]
> @Otterverse just for more context - the issue we've been seeing is that some images of raspian bullseye seem to install `libpigpio1=1.79-1+rpt1.1 libpigpio-dev=1.79-1+rpt1.1`, which seem to not adhere to what libgpiod1 is compatible with.
> 
> ```shell
> The following information may help to resolve the situation:
> 
> The following packages have unmet dependencies:
>  pigpio : Depends: libpigpio1 (= 1.79-1+rpt1) but 1.79-1+rpt1.1 is to be installed
>           Depends: libpigpio-dev (= 1.79-1+rpt1) but 1.79-1+rpt1.1 is to be installed
> E: Unable to correct problems, you have held broken packages.
> ```
> 
> This was an issue fixed for bookworm with pigpio, which Susmita has more context on, but I could not get the module to run without erroring on bullseye with the dynamic link without forcing `rpt1` on some bullseye raspberry pi4s with `sudo apt install libpigpio1=1.79-1+rpt1 libpigpio-dev=1.79-1+rpt1`. This has popped up on 3 separate raspberry pis since the module's release, and I'm sure there are more.

Okay, that helps for context. I'll try to repro on a bullseye pi tomorrow. It's very strange to me that bullseye reported having a NEWER version available. Maybe they corrupted the repo temporarily and had to rollback (which is supposed to "never" happen)? The proper fix is likely to remove and reinstall ALL pigpio packages though. Doing just two risks leaving other stuff in a mismatched state.

**Comment by Otterverse** [comment]
Actually, I just figured it out, and it's a giant facepalm!

We (specifically ME) made the rpt1.1 packages ages ago, and put them in our PRIVATE viam repo. I only did this explicitly for the dev environment we previously needed (e.g. just the library and dev files, no pigpiod or other related libraries.) So we could build with that library on other machines (and docker images) that aren't raspberry pi themselves. So the pis in question must've been set up with our private repo (probably someone ran `make setup` in RDK on them, or they were the REALLY old custom viam pi images.)

As such, this should NOT be a problem for the wider world. But definitely breaks things after having run rdk's setup script on. I don't think it's worth trying to "work around" it in the module itself. We should just fix the pis (and remove the installation of the pi libraries from RDK's setup.sh (and then remove them from our repo as well.)

 @susmitaSanyal @randhid you may both feel free to yell at me now!  :-)

**Comment by randhid** [comment]
> Actually, I just figured it out, and it's a giant facepalm!
> 
> We (specifically ME) made the rpt1.1 packages ages ago, and put them in our PRIVATE viam repo. I only did this explicitly for the dev environment we previously needed (e.g. just the library and dev files, no pigpiod or other related libraries.) So we could build with that library on other machines (and docker images) that aren't raspberry pi themselves. So the pis in question must've been set up with our private repo (probably someone ran `make setup` in RDK on them, or they were the REALLY old custom viam pi images.)
> 
> As such, this should NOT be a problem for the wider world. But definitely breaks things after having run rdk's setup script on. I don't think it's worth trying to "work around" it in the module itself. We should just fix the pis (and remove the installation of the pi libraries from RDK's setup.sh (and then remove them from our repo as well.)
> 
> @susmitaSanyal @randhid you may both feel free to yell at me now! :-)

It's me, hi, I'm the problem it's me.

**Review by randhid — APPROVED** [review]
Approving to unblock.

**Comment by susmitaSanyal** [comment]
@Otterverse So is this good to go?

**Comment by Otterverse** [comment]
> @Otterverse So is this good to go?

Sorry, no... As I said in my last post, this shouldn't try to workaround a mistake in our internal use stuff. I really think it should just `apt-get install -qqy pigpio` and call it a day. Pinning the other two versions just means potential future breakage if/when upstream updates the pigpio version.

**Comment by randhid** [comment]
> > @Otterverse So is this good to go?
> 
> Sorry, no... As I said in my last post, this shouldn't try to workaround a mistake in our internal use stuff. I really think it should just `apt-get install -qqy pigpio` and call it a day. Pinning the other two versions just means potential future breakage if/when upstream updates the pigpio version.

ie - let's go back to the simpler version that we know works on a freshly flashed bullseye. We will ask our coworkers to run that command if they have a pi that uses jame's magic rpt1.1 packages.

**Comment by susmitaSanyal** [comment]
@Otterverse @randhid alright, keeping it simple and closing this PR.

**Comment by npentrel** [comment]
I had this issue and ran the above commands which allowed the module to start. However I've noticed that I am also having issues with the servo component:
```
11/7/2024, 12:53:26 PM error rdk.resource_manager.rdk:component:servo/servo resource/graph_node.go:296 resource build error: unknown resource type: API "rdk:component:servo" with model "viam:raspberry-pi:pi-servo" not registered resource rdk:component:servo/servo model viam:raspberry-pi:pi-servo
```

Any thoughts on whether that's related?

**Pull request closed without merging**